### PR TITLE
Added tile endpoint, S3 caching mechanism and various fixes and improvements

### LINKED ISSUE #97: Add endpoint for tiling
In order to display orthophotos via https://github.com/DroneDB/CommonUI/blob/master/classes/olHybridXYZ.js we need to create a TMS-style endpoint. 

Tiling is already available via `ddb tile`.

Bindings will need to be created (both JS and .NET)

### PR DESCRIPTION
**Pull request by HeDo88TH**
Closes #97 

Tested with orthophoto https://github.com/DroneDB/test_data/blob/master/brighton/odm_orthophoto.tif

### CODE REVIEW

File: Registry.Web/Services/Adapters/ObjectsManager.cs [hunk 1]
```diff
@@ -388,12 +383,58 @@ public async Task<FileDescriptorDto> GenerateThumbnail(string orgSlug, string ds
...
+                throw new ArgumentException("Path is not valid");
+
+            var destFilePath = string.Empty;
+            var sourceFilePath = Path.Combine(Path.GetTempPath(), "out-" + fileName);
+
+            try
+            {
+
+                var obj = await InternalGet(orgSlug, ds.InternalRef, path);
+                await File.WriteAllBytesAsync(sourceFilePath, obj.Data);
```
Comment (pierotofy) [inline Registry.Web/Services/Adapters/ObjectsManager.cs:414]: What's the logic for making a copy of the file?

These files could be large (1GB+). Is this getting called at each tile request?

Comment (HeDo88TH) [reply Registry.Web/Services/Adapters/ObjectsManager.cs:414]: Addressed by https://github.com/DroneDB/Registry/pull/109/commits/9ad5fb10b07af17124f0ca8ffcec0ca31f7de633

**Comment by HeDo88TH** [comment]
Improved #102 and #103 . The cache cleanup logic is still work in progress. But now I inserted a caching proxy class (CachedS3ObjectSystem) between S3 and ObjectManager in order to avoid remote storage calls of already downloaded data. Plus, when performing share all the files are pre-cached and stored locally. This way they are ready to be shown to the user.
The new file caching system still needs work on file expiration and size constraints but I think that It is the right direction.

**Comment by HeDo88TH** [comment]
Closes #103 
When files are in cache they are read from the disk in order to avoid S3 latency. The cache can be limited in size and it supports file expiration calculated on last access.

Closes #102 
When uploading files, Registry moves the S3 upload to a different thread and returns immediately. This eliminates the delay and makes the upload go much faster. The drawback is that there is no way (not yet) to know when the dataset is fully uploaded, only browsing it gives a clue. BTW when file sizes are contained there is not virtually any problem. If the last file is 8GB we can start preoccupaying because we should add a flag to the dataset in order to signal when the upload is finished.

I will push these changes to testhub when the PR is approved

**Comment by HeDo88TH** [comment]
In order to test the cache you can use this settings:
Local minio started with
`minio server ./temp`

```
   "StorageProvider": {
      "Type": "CachedS3",
      "Settings": {
        "Endpoint": "localhost:9000",
        "AccessKey": "minioadmin",
        "SecretKey": "minioadmin",
        "UseSsl": false,
        "AppName": "Registry",
        "AppVersion": "1.0",
        "Region": "us-east-1",
        "CachePath": "C:\\cache",
        "MaxSize": 52428800,
        "CacheExpiration": "00:15:00"
      }
    },
```

Or to wasabi

    ```
"StorageProvider": {
      "Type": "CachedS3",
      "Settings": {
        "Endpoint": "s3.wasabisys.com",
        "AccessKey": "XX",
        "SecretKey": "XX",
        "UseSsl": false,
        "AppName": "Registry",
        "AppVersion": "1.0",
        "Region": "us-east-1",
        "CachePath": "C:\\cache",
        "MaxSize": 209715200,
        "CacheExpiration": "00:15:00"
      }
    },
```

`MaxSize` is in bytes and `CachePath` can be an absolute or relative path

**Comment by pierotofy** [comment]
Will the cache be added into the docker-compose?

**Comment by HeDo88TH** [comment]
> Will the cache be added into the docker-compose?

Yes

**Comment by pierotofy** [comment]
Nice performance increase! :+1: 

Found a pretty big issue though:

- Setup with local minio cache, sqlite
- I upload two datasets:

```
[piero:/data … righton2/images] [.venv] $ ddb share DJI_0019.JPG -s http://192.168.2.251:5000
DJI_0019.JPG [###########################] 100.00% 5.67s
http://192.168.2.251:5000/r/admin/bknndotyqfivorn9
[piero:/data … righton2/images] [.venv] $ ddb share DJI_0018.JPG -s http://192.168.2.251:5000
DJI_0018.JPG [###########################] 100.00% 5.37s
http://192.168.2.251:5000/r/admin/igxwn1agj4zvbb9n
```

image

igxwn1agj4zvbb9n --> DJI_0018.JPG
bknndotyqfivorn9 --> DJI_0019.JPG

- Both dataset open fine.
- I delete `bknndotyqfivorn9` (first dataset, DJI_0019.JPG)
- I open `igxwn1agj4zvbb9n`

Files don't show up (`http://localhost:5000/orgs/admin/ds/igxwn1agj4zvbb9n/list` returns empty list). It opened fine before deleting `bknndotyqfivorn9` (the other dataset).

**Comment by HeDo88TH** [comment]
Closes #111 
It was an inconsistency in configuration: connectionstring was pointing to sqlite but mysql was specified

**Comment by pierotofy** [comment]
Screenshot from another run (different datasetID) shows the call to `/list` takes a while, but eventually returns empty list.

image

No error messages in the log, no stack traces.

**Comment by HeDo88TH** [comment]
> Screenshot from another run (different datasetID) shows the call to `/list` takes a while, but eventually returns empty list.
> 
> image
> 
> No error messages in the log, no stack traces.

I will look into this issue as soon as possible

**Comment by HeDo88TH** [comment]
Fixed issue (https://github.com/DroneDB/Registry/pull/109#issuecomment-762383983 and https://github.com/DroneDB/Registry/pull/109#issuecomment-762384554)

**Comment by HeDo88TH** [comment]
Also closes #112

File: Registry.Web/Services/Ports/ISystemManager.cs [hunk 2]
```diff
@@ -2,13 +2,17 @@
...
 
 namespace Registry.Web.Services.Ports
 {
     public interface ISystemManager
     {
         public Task<CleanupResult> CleanupSessions();
-        Task SyncDdbMeta(string[] orgs = null, bool skipAuthCheck = false);
+        public Task SyncDdbMeta(string[] orgs = null, bool skipAuthCheck = false);
+
+        // This is a hard reference and a violation of our architecture. Keeping it simple by now.
```
Comment (pierotofy) [inline Registry.Web/Services/Ports/ISystemManager.cs:15]: Simple is good!

Comment (HeDo88TH) [reply Registry.Web/Services/Ports/ISystemManager.cs:15]: There are still some rough edges that I need to fix. This is one of those 👍

Comment (HeDo88TH) [reply Registry.Web/Services/Ports/ISystemManager.cs:15]: Fixed in https://github.com/DroneDB/Registry/pull/109/commits/6e2ba75b7c225344663020dd5c9ca8cf9387c7d2

**Comment by HeDo88TH** [comment]
Closes #110 but we need a more standardized way to address the auth scheme

**Comment by HeDo88TH** [comment]
It is ready to test

File: Registry.Web/appsettings.json [hunk 3]
```diff
@@ -3,8 +3,8 @@
     "Secret": "a2780070a24cfcaf5a4a43f931200ba0d19d8b86b3a7bd5123d9ad75b125f480fcce1f9b7f41a53abe2ba8456bd142d38c455302e0081e5139bc3fc9bf614497",
     "TokenExpirationInDays": 7,
     "RevokedTokens": [],
-    "AuthProvider": "Sqlite",
-    "RegistryProvider": "Sqlite",
+    "AuthProvider": "Mysql",
```
Comment (pierotofy) [inline Registry.Web/appsettings.json]: Was this committed by mistake?

Comment (HeDo88TH) [reply Registry.Web/appsettings.json]: Definitely but is now fixed by https://github.com/DroneDB/Registry/pull/109/commits/9464ff41cd0d481ee784da2638db4f9fc7f529af

**Comment by HeDo88TH** [comment]
New endpoint `/sys/syncfiles` that checks for failed uploads and retries them. 
TIP: In order to make uploads fail it is sufficient to add 
throw new Exception("dummy") before line https://github.com/DroneDB/Registry/pull/109/files#diff-89822cd7e9464c34a81cfb1fe187151dbe380f7167bad71a5bb1143b42710a4fR666

**Comment by pierotofy** [comment]
> line 666 

coincidence? lol

**Comment by pierotofy** [comment]
Basic tests seem to work as expected. I've added some comments in #114 related to the design, but can be handled as part of a separate issue.

**Pull request merged**